intro-to-dev-environment/intro-to-computer-memory.md

```diff
@@ -0,0 +1,166 @@
+
+We could say that in this region, self-storage units are assigned whenever someone comes in and asks for a certain number of storage units. The self-storage unit manager will do their best to give the customer a row of storage units all next to each other. When a renter is finished with their storage unit and is done with it, the manager could consider re-renting that unit again. However, navigating the storage units here may be trickier, or just take a little longer.
+
+### !end-callout
+
+### Stack and Heap Summary
+
+| Region | How it stores data                                                                                                                                                       | Strength                                                 |
+| ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------------------------------------- |
+| Stack  | Like a stack of dishes. New data gets placed "on top" of the stack of dishes. The data that is the last to come into the stack is the data that's the first to come out. | _Very_ fast for occasions where data is _very_ temporary |
+| Heap   | Determines how much and where memory is allocated as data of varying sizes comes in.                                                                                     | Stores data of unknown sizes well                        |
+
+## Applied to Variables
+
+Consider variable assignment:
+
+```python
+apples = "Hello, World!"
+```
+
+When this program runs, our CPU begins to process it.
+
+Every time our CPU runs into variable assignment, our CPU wants to put that data in memory. However, the CPU doesn't want to store the full idea "the variable `apples` has the value `"Hello, World!"`. Instead, the CPU thinks of **_two things_** that should be held in memory:
+
+1. On the heap, store the value of the variable at a specific memory address
+2. On the stack, store the variable and the specific memory address that it _points to_
+
+### Example
+
+When running the line `apples = "Hello, World!"`, the CPU wants to store these two things in memory:
+
+1. The value `"Hello, World!"`
+2. The variable `apples`
+
+The CPU decides to process `apples = "Hello, World!"` by allocating memory like this:
+
+1. On the heap, the string value `"Hello, World!"` gets stored with a memory address of `5555`
+2. On the stack, the variable named `apples` gets stored. Along with the variable name, it stores the information "the value that this variable points to is in the memory address `5555`."
+
+## Summary
+
+This primer on computer memory should help us think more like a computer, and empathize with how a computer stores data just a little bit more.
+
+We can also feel a closer relationship between the physical computer hardware, the data in memory, and the variables we work with in Python code.
```

Comment:
Computers don't think. They're dumb. They have no feelings, so we can't empathize with them. If we feel a closer relationship between those things, why should that matter? Let's be more clear about what kinds of benefits the understanding the relationships can bring.

In the Summary, we should summarize what was covered, and what the takeaways are.

```suggestion
The most important components in computers are the CPU and RAM. They work together to run our programs, with the CPU processing instructions and data that are stored in RAM. RAM usage must be managed carefully, and two strategies of doing so are as a stack, and as a heap. Python variables store the value and the variable name as two different pieces of information, and take advantage of the strengths of stack management and heap management to store the value on the heap and the name on the stack, along with a pointer to the heap value.

Understanding how a computer runs our programs, and awareness of how Python tracks data in memory, can give us a more concrete idea of how to structure our programs and data to make the best use of the computer's resources.
```